core/network/fabric-interop-cc/contracts/interop/membership.go

```diff
@@ -9,45 +9,334 @@
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal attested membership set: %s", err.Error())
+	}
+	decodedForeignMembership, err := base64.StdEncoding.DecodeString(attestedMembershipSet.Membership)
+	if err != nil {
+		return fmt.Errorf("Foreign membership could not be decoded from base64: %s", err.Error())
+	}
+	var foreignMembership common.Membership
+	err = protoV2.Unmarshal(decodedForeignMembership, &foreignMembership)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal membership: %s", err.Error())
+	}
+
+	// Match nonces across all attestations, local and foreign
+	matchedNonce := ""
+	for _, attestation := range append(counterAttestedMembership.Attestations, attestedMembershipSet.Attestations...) {
+		if matchedNonce == "" {
+			matchedNonce = attestation.Nonce
+		} else {
+			if matchedNonce != attestation.Nonce {
+				return fmt.Errorf("Mismatched nonces across two attestations: %s, %s", matchedNonce, attestation.Nonce)
+			}
+		}
+	}
+
+	// Ensure valid attestations from all local IIN Agents
+	for _, attestation := range counterAttestedMembership.Attestations {
+		err = validateAttestation(attestation, counterAttestedMembership.AttestedMembershipSet + attestation.Nonce)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Validate foreign membership cert chains
+	err = validateMemberCertChains(&foreignMembership)
+	if err != nil {
+		return err
+	}
+
+	// Ensure authentic and valid attestations from all foreign IIN Agents
+	for _, attestation := range attestedMembershipSet.Attestations {
+		if attestation.UnitIdentity.SecurityDomain != foreignMembership.SecurityDomain {
+			return fmt.Errorf("Foreign agent security domain %s does not match attested membership security domain %s",
+				attestation.UnitIdentity.SecurityDomain, foreignMembership.SecurityDomain)
+		}
+		foundMemberMatch := false
+		for _, member := range foreignMembership.Members {
+			var isSignerRoot bool
+			var leafCertPEM string
+			if member.Type == "ca" {
+				leafCertPEM = member.Value
+				isSignerRoot = true
+			} else if member.Type == "certificate" {
+				leafCertPEM = member.Chain[len(member.Chain) - 1]
+				isSignerRoot = (len(member.Chain) == 1)
+			}
+			attesterCert, err := parseCert(attestation.Certificate)
+			if err != nil {
+				continue
+			}
+			leafCert, err := parseCert(leafCertPEM)
+			if err != nil {
+				continue
+			}
+			err = validateCertificateUsingCA(attesterCert, leafCert, isSignerRoot)
+			if err == nil {
+				foundMemberMatch = true
+				break
+			}
+		}
+		if !foundMemberMatch {
+			return fmt.Errorf("No matching member certificate chain found for attester")
+		}
+		// Validate signature
+		err = validateAttestation(attestation, attestedMembershipSet.Membership + attestation.Nonce)
+		if err != nil {
+			return err
+		}
+	}
+
+	membershipKey, err := ctx.GetStub().CreateCompositeKey(membershipObjectType, []string{foreignMembership.SecurityDomain})
```

Comment:
I feel this check should happen first, at the top of this function before anything else.

Reply:
Fixed here and in the update function.